### ISSUE TITLE
Non-deterministic NaN corruption during CPU→GPU tensor transfer on gfx1030 (RX 6900 XT)

### ISSUE THREAD

**Issue by almara3** [RadeonOpenCompute/ROCm#6123]
**System Information**
GPU: AMD Radeon RX 6900 XT (gfx1030)
ROCm: 6.4.2
PyTorch: 2.9.1+rocm6.4
OS: Linux Mint 21.2 (Ubuntu 22.04 Jammy)
Kernel: Linux 5.15.0-153-generic
Driver: amdgpu

**Problem Description**
Non-deterministic NaN corruption occurs when transferring tensors from CPU to GPU using .to("cuda"). The corruption is not caused by the tensor values themselves.

**Minimal Reproducer**
```
import torch

# Create tensor once on CPU - data is constant, not recreated
w = torch.randn(13000, 384)

# Repeat only the transfer
for run in range(10):
    w_gpu = w.to("cuda")
    print(f"Run {run}: NaN={torch.isnan(w_gpu).any()}")

# Expected: all False
# Actual: NaN appears non-deterministically, more frequently after the first transfer
```

**Observations**
Bug affects tensors of all sizes
The first or second transfer of a new tensor succeeds more often than subsequent transfers
Repeated transfers of the same tensor fail more frequently
The bug is non-deterministic but highly reproducible over multiple runs
Random data and real model weights are both affected

**Real-world Impact**
This bug breaks PyTorch model inference. For example, OpenAI Whisper's token_embedding.weight matrix (51865 × 384, float32) is silently corrupted when loaded onto the GPU, producing NaN logits and crashing inference. The model loads without error but produces invalid results.

**Workaround**
Retry the transfer until no NaN values are detected:
pythonfor attempt in range(10):
    w_gpu = w_cpu.to("cuda")
    if not torch.isnan(w_gpu).any():
        break  # Transfer successful
This typically succeeds within 1–2 attempts.

**Additional Notes**
- torch.cuda.is_available() returns True and the GPU is correctly detected
- rocminfo correctly identifies the GPU as gfx1030
- No error or warning is raised during the corrupted transfer – the corruption is silent

**Comment by almara3** [comment]
### Update: Fixed in kernel 6.8

After upgrading from kernel 5.15.0-153-generic to 6.8.0-107-generic (Ubuntu HWE), the bug seems to be gone.

**Comment by Qubitium** [comment]
@almara3  Are you using the `amdgpu-dkms` driver or the one that came with the kernel?

**Comment by 1fanwang** [comment]
Two suggestions that might help, given @almara3's data point that 6.8 fixes it:

**Narrowing the driver source for @Qubitium's question.** On a 5.15 box, both paths are plausible — Ubuntu 22.04 ships the in-kernel `amdgpu` driver, but ROCm's `amdgpu-dkms` package replaces it once installed and reuses the same module name. These should disambiguate:

```bash
modinfo amdgpu | grep -E '^(filename|version|srcversion|vermagic)'
dkms status amdgpu 2>/dev/null
apt list --installed 2>/dev/null | grep amdgpu-dkms
```

If @almara3 was on the in-kernel `amdgpu` and the upgrade to 6.8 picked up the in-kernel one too, the fix likely lives in the upstream amdgpu DMA-BUF / TTM handling that landed in the 6.x window. If they were on `amdgpu-dkms`, the dkms tarball may already carry the fix backported to 5.15.

**A more targeted workaround to test on 5.15.** The retry loop works for "load weights once" but doesn't help streaming inference paths. Pinned-memory transfer with explicit sync uses a different DMA route than the default pageable copy and may avoid the corruption hot-path on 5.15:

```python
w_pinned = w.pin_memory()
w_gpu = w_pinned.to("cuda", non_blocking=False)
torch.cuda.synchronize()
```

This is a hypothesis worth verifying on 5.15 + RX 6900 XT — if it eliminates the corruption, it gives users a single-shot workaround while they migrate to 6.8, and confirms the bounce-buffer / pageable copy path as the corruption source. Happy to dig into the amdgpu kernel changes between 5.15 and 6.8 if it would help track down the specific upstream commit, since that would inform whether `amdgpu-dkms` users on 5.15 are still affected.

**Comment by almara3** [comment]
@Qubitium On kernel 5.15.0-153-generic I was using the amdgpu-dkms driver (version 5.18.13-1577590.22.04, installed via ROCm). After upgrading to kernel 6.8.0-107-generic (Ubuntu HWE), the dkms module was broken (source files missing), so the inbox amdgpu driver from kernel 6.8 is now used instead.

Edit: Formatting
